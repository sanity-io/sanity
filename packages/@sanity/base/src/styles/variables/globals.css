:root {
  /* Colors */
  --body-bg: #fff;
  --body-text: color(var(--body-bg) contrast(80%));
  --component-bg: var(--body-bg);
  --component-text-color: color(var(--component-bg) contrast(80%));
  --preview-placeholder-color: color(var(--component-bg) contrast(80%) alpha(10%));
  --component-border-color: color(var(--component-bg) contrast(80%) a(20%));
  --backdrop-color: color(var(--gray-base) a(20%));
  --main-navigation-color: color(var(--brand-primary) shade(70%) saturation(20%));
  --main-navigation-color--inverted: color(var(--main-navigation-color) contrast(80%));
  --dialog-header-color: color(var(--brand-primary) shade(20%) saturation(30%));
  --dialog-header-color--inverted: var(--main-navigation-color--inverted);
  --header-height: 3rem;
  --checkerboard-color: color(var(--component-text-color) a(5%));
  --selectable-item-base: #156dff;
  --default-button-primary-color: #156dff;
  --default-button-primary-color--inverted: #fff;

  /* Shadows */
  --shadow-base: color(var(--gray-base) shade(30%));
  --shadow-color-umbra: color(var(--shadow-base) a(2%));
  --shadow-color-penumbra: color(var(--shadow-base) a(14%));
  --shadow-color-ambient: color(var(--shadow-base) a(12%));

  /* Border radius */
  --border-radius-base: 4px;
  --border-radius-large: 6px;
  --border-radius-medium: var(--border-radius-base);
  --border-radius-small: 2px;

  /* Padding */
  --extra-small-padding: 0.25rem;
  --small-padding: 0.5rem;
  --medium-padding: 1rem;
  --large-padding: 1.5rem;
  --extra-large-padding: 3rem;
  --text-selection-color: color(var(--brand-primary) alpha(20%));
}
