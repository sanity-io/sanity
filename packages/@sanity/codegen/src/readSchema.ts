import {readFile} from 'node:fs/promises'

import {type SchemaType} from 'groq-js'

/**
 * Read a schema from a given path
 * @param path - The path to the schema
 * @returns The schema
 * @internal
 * @beta
 **/
export async function readSchema(path: string): Promise<SchemaType> {
  const content = await readFile(path, 'utf-8')
  return JSON.parse(content) // todo: ZOD validation?
}
