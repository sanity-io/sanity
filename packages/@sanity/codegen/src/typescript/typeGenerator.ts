import {CodeGenerator} from '@babel/generator'
import * as t from '@babel/types'
import {
  type ArrayTypeNode,
  type DocumentSchemaType,
  type ObjectAttribute,
  type ObjectTypeNode,
  type SchemaType,
  type TypeDeclarationSchemaType,
  type TypeNode,
  type UnionTypeNode,
} from 'groq-js'

const REFERENCE_SYMBOL_NAME = 'internalGroqTypeReferenceTo'

/**
 * A class used to generate TypeScript types from a given schema
 * @internal
 * @beta
 */
export class TypeGenerator {
  private generatedTypeName: Set<string> = new Set()
  private typeNameMap: Map<string, string> = new Map()

  private readonly schema: SchemaType

  constructor(schema: SchemaType) {
    this.schema = schema
  }

  /**
   * Generate TypeScript types for the given schema
   * @returns string
   * @internal
   * @beta
   */
  generateSchemaTypes(): string {
    const typeDeclarations: (t.TSTypeAliasDeclaration | t.ExportNamedDeclaration)[] = []

    this.schema.forEach((schema) => {
      const typeLiteral = this.getTypeNodeType(schema)

      const typeAlias = t.tsTypeAliasDeclaration(
        t.identifier(this.getTypeName(schema.name)),
        null,
        typeLiteral,
      )

      typeDeclarations.push(t.exportNamedDeclaration(typeAlias))
    })

    // Generate TypeScript code from the AST nodes
    return typeDeclarations.map((decl) => new CodeGenerator(decl).generate().code).join('\n\n')
  }

  /**
   * Takes a identifier and a type node and generates a type alias for the type node.
   * @param identifierName - The name of the type to generated
   * @param typeNode - The type node to generate the type for
   * @returns
   * @internal
   * @beta
   */
  generateTypeNodeTypes(identifierName: string, typeNode: TypeNode): string {
    const type = this.getTypeNodeType(typeNode)

    const typeAlias = t.tsTypeAliasDeclaration(
      t.identifier(this.getTypeName(identifierName)),
      null,
      type,
    )

    return new CodeGenerator(t.exportNamedDeclaration(typeAlias)).generate().code
  }

  static generateKnownTypes(): string {
    const typeOperator = t.tsTypeOperator(t.tsSymbolKeyword())
    typeOperator.operator = 'unique'

    const identifier = t.identifier(REFERENCE_SYMBOL_NAME)
    identifier.typeAnnotation = t.tsTypeAnnotation(typeOperator)

    const decleration = t.variableDeclaration('const', [t.variableDeclarator(identifier)])
    decleration.declare = true
    return new CodeGenerator(t.exportNamedDeclaration(decleration)).generate().code
  }

  /**
   * Since we are sanitizing identifiers we migt end up with collisions. Ie there might be a type mux.video and muxVideo, both these
   * types would be sanityized into MuxVideo. To avoid this we keep track of the generated type names and add a index to the name.
   * When we reference a type we also keep track of the original name so we can reference the correct type later.
   */
  private getTypeName(name: string): string {
    const desiredName = uppercaseFirstLetter(sanitizeIdentifier(name))

    let generatedName = desiredName
    let i = 2
    while (this.generatedTypeName.has(generatedName)) {
      // add _ and a index and increment that index until we find a name that is not in the map
      generatedName = `${desiredName}_${i++}`
    }
    this.generatedTypeName.add(generatedName)
    this.typeNameMap.set(name, generatedName)
    return generatedName
  }

  private getTypeNodeType(
    typeNode: TypeNode | TypeDeclarationSchemaType | DocumentSchemaType,
  ): t.TSType {
    switch (typeNode.type) {
      case 'string': {
        if (typeNode.value !== undefined) {
          return t.tsLiteralType(t.stringLiteral(typeNode.value))
        }
        return t.tsStringKeyword()
      }
      case 'number': {
        if (typeNode.value !== undefined) {
          return t.tsLiteralType(t.numericLiteral(typeNode.value))
        }
        return t.tsNumberKeyword()
      }
      case 'boolean': {
        if (typeNode.value !== undefined) {
          return t.tsLiteralType(t.booleanLiteral(typeNode.value))
        }
        return t.tsBooleanKeyword()
      }
      case 'unknown': {
        return t.tsUnknownKeyword()
      }
      case 'document': {
        return this.generateDocumentType(typeNode)
      }
      case 'type': {
        return this.getTypeNodeType(typeNode.value)
      }
      case 'array': {
        return this.generateArrayTsType(typeNode)
      }
      case 'object': {
        return this.generateObjectTsType(typeNode)
      }
      case 'union': {
        return this.generateUnionTsType(typeNode)
      }
      case 'inline': {
        return t.tsTypeReference(
          t.identifier(uppercaseFirstLetter(sanitizeIdentifier(typeNode.name))),
        )
      }
      case 'null': {
        return t.tsNullKeyword()
      }

      default:
        // @ts-expect-error This should never happen
        throw new Error(`Type "${typeNode.type}" not found in schema`)
    }
  }

  // Helper function used to generate TS types for array type nodes.
  private generateArrayTsType(typeNode: ArrayTypeNode): t.TSTypeReference {
    const typeNodes = this.getTypeNodeType(typeNode.of)
    const arrayType = t.tsTypeReference(
      t.identifier('Array'),
      t.tsTypeParameterInstantiation([typeNodes]),
    )

    return arrayType
  }

  // Helper function used to generate TS types for object properties.
  private generateObjectProperty(key: string, attribute: ObjectAttribute): t.TSPropertySignature {
    const type = this.getTypeNodeType(attribute.value)
    const propertySignature = t.tsPropertySignature(
      t.identifier(sanitizeIdentifier(key)),
      t.tsTypeAnnotation(type),
    )
    propertySignature.optional = attribute.optional

    return propertySignature
  }

  // Helper function used to generate TS types for object type nodes.
  private generateObjectTsType(typeNode: ObjectTypeNode): t.TSType {
    const props: t.TSPropertySignature[] = []
    Object.entries(typeNode.attributes).forEach(([key, attribute]) => {
      props.push(this.generateObjectProperty(key, attribute))
    })
    if (typeNode.rest !== undefined) {
      switch (typeNode.rest.type) {
        case 'unknown': {
          return t.tsUnknownKeyword()
        }
        case 'object': {
          Object.entries(typeNode.rest.attributes).forEach(([key, attribute]) => {
            props.push(this.generateObjectProperty(key, attribute))
          })
          break
        }
        case 'inline': {
          return t.tsIntersectionType([
            t.tsTypeLiteral(props),
            t.tsTypeReference(
              t.identifier(
                this.typeNameMap.get(typeNode.rest.name) ||
                  uppercaseFirstLetter(sanitizeIdentifier(typeNode.rest.name)),
              ),
            ),
          ])
        }
        default: {
          // @ts-expect-error This should never happen
          throw new Error(`Type "${typeNode.rest.type}" not found in schema`)
        }
      }
    }
    if (typeNode.dereferencesTo !== undefined) {
      const derefType = t.tsPropertySignature(
        t.identifier(REFERENCE_SYMBOL_NAME),
        t.tsTypeAnnotation(t.tsLiteralType(t.stringLiteral(typeNode.dereferencesTo))),
      )
      derefType.computed = true
      derefType.optional = true
      props.push(derefType)
    }
    return t.tsTypeLiteral(props)
  }

  // Helper function used to generate TS types for union type nodes.
  private generateUnionTsType(typeNode: UnionTypeNode): t.TSType {
    if (typeNode.of.length === 0) {
      return t.tsNeverKeyword()
    }
    if (typeNode.of.length === 1) {
      return this.getTypeNodeType(typeNode.of[0])
    }

    const typeNodes = typeNode.of.map((node) => this.getTypeNodeType(node))

    return t.tsUnionType(typeNodes)
  }

  // Helper function used to generate TS types for document type nodes.
  private generateDocumentType(document: DocumentSchemaType): t.TSType {
    const props = Object.entries(document.attributes).map(([key, node]) =>
      this.generateObjectProperty(key, node),
    )

    return t.tsTypeLiteral(props)
  }
}
function uppercaseFirstLetter(input: string): string {
  return input.charAt(0).toUpperCase() + input.slice(1)
}

function sanitizeIdentifier(input: string): string {
  return `${input.replace(/^\d/, '_').replace(/[^$\w]+(.)/g, (_, char) => char.toUpperCase())}`
}
