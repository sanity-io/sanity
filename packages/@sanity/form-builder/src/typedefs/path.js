
type KeyedSegment = {
  _key: string
}

export type PathSegment = string | number | KeyedSegment

export type Path = Array<PathSegment>
