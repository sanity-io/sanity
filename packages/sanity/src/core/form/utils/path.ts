import {arrayToJSONMatchPath} from '@sanity/mutator'
import {Path, PathSegment} from '@sanity/types'

const IS_NUMERIC = /^\d+$/

function unquote(str: string) {
  return str.replace(/^['"]/, '').replace(/['"]$/, '')
}

function splitAttr(segment: string) {
  const [attr, key] = segment.split('==')
  return {[attr]: unquote(key)}
}

function coerce(segment: string): PathSegment {
  return IS_NUMERIC.test(segment) ? Number(segment) : segment
}

function parseGradientPath(focusPathStr: string): Path {
  return focusPathStr
    .split(/[[.\]]/g)
    .filter(Boolean)
    .map((seg) => (seg.includes('==') ? splitAttr(seg) : coerce(seg))) as Path
}

/**
 * @internal
 */
export function encodePath(formBuilderPath: Path): string {
  return arrayToJSONMatchPath(formBuilderPath)
}

/**
 * @internal
 */
export function decodePath(gradientPath: string): Path {
  return parseGradientPath(gradientPath)
}
