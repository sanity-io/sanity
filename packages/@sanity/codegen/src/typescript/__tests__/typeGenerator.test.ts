import path from 'node:path'

import {describe, expect, test} from '@jest/globals'
import {
  createReferenceTypeNode,
  type DocumentSchemaType,
  type ObjectAttribute,
  type ObjectTypeNode,
  type SchemaType,
  type StringTypeNode,
  type TypeNode,
} from 'groq-js'

import {readSchema} from '../../readSchema'
import {TypeGenerator} from '../typeGenerator'

describe('generateSchemaTypes', () => {
  test('should generate TypeScript type declarations for a schema', async () => {
    const schema = await readSchema(path.join(__dirname, 'fixtures', 'schema.json'))

    const typeGenerator = new TypeGenerator(schema)
    const typeDeclarations = typeGenerator.generateSchemaTypes()

    expect(typeDeclarations).toMatchSnapshot()
  })

  test('can generate well known types', async () => {
    const typeDeclarations = TypeGenerator.generateKnownTypes()

    expect(typeDeclarations).toMatchSnapshot()
  })

  test('should generate correct types for document schema with string fields', () => {
    const schema: SchemaType = [
      {
        type: 'document',
        name: 'author',
        attributes: {
          _id: {
            type: 'objectAttribute',
            value: {type: 'string'},
          },
          name: {
            type: 'objectAttribute',
            value: {type: 'string'},
            optional: true,
          },
        },
      },
    ]

    const typeGenerator = new TypeGenerator(schema)
    const actualOutput = typeGenerator.generateSchemaTypes()

    expect(actualOutput).toMatchInlineSnapshot(`
"export type Author = {
  _id: string;
  name?: string;
};

export type AllSanitySchemaTypes = Author;"
`)
  })

  test('should generate correct types for document schema with number fields', () => {
    const schema: SchemaType = [
      {
        type: 'document',
        name: 'product',
        attributes: {
          price: {
            type: 'objectAttribute',
            value: {type: 'number'},
          },
        },
      },
    ]

    const typeGenerator = new TypeGenerator(schema)
    const actualOutput = typeGenerator.generateSchemaTypes()

    expect(actualOutput).toMatchInlineSnapshot(`
"export type Product = {
  price: number;
};

export type AllSanitySchemaTypes = Product;"
`)
  })

  test('should generate correct types for document schema with boolean fields', () => {
    const schema: SchemaType = [
      {
        type: 'document',
        name: 'task',
        attributes: {
          completed: {
            type: 'objectAttribute',
            value: {type: 'boolean'},
          },
        },
      },
    ]
    const typeGenerator = new TypeGenerator(schema)
    const actualOutput = typeGenerator.generateSchemaTypes()

    expect(actualOutput).toMatchInlineSnapshot(`
"export type Task = {
  completed: boolean;
};

export type AllSanitySchemaTypes = Task;"
`)
  })

  test('should generate correct types for document schema with object fields', () => {
    const schema: SchemaType = [
      {
        type: 'document',
        name: 'user',
        attributes: {
          address: {
            type: 'objectAttribute',
            value: {
              type: 'object',
              attributes: {
                street: {
                  type: 'objectAttribute',
                  value: {type: 'string'},
                },
                city: {
                  type: 'objectAttribute',
                  value: {type: 'string'},
                },
              },
            },
          },
        },
      },
    ]

    const typeGenerator = new TypeGenerator(schema)
    const actualOutput = typeGenerator.generateSchemaTypes()

    expect(actualOutput).toMatchInlineSnapshot(`
"export type User = {
  address: {
    street: string;
    city: string;
  };
};

export type AllSanitySchemaTypes = User;"
`)
  })

  test('should generate correct types for document schema with array fields', () => {
    const schema: SchemaType = [
      {
        type: 'document',
        name: 'blogPost',
        attributes: {
          tags: {
            type: 'objectAttribute',
            value: {
              type: 'array',
              of: {type: 'string'},
            },
          },
        },
      },
    ]

    const typeGenerator = new TypeGenerator(schema)
    const actualOutput = typeGenerator.generateSchemaTypes()

    expect(actualOutput).toMatchInlineSnapshot(`
"export type BlogPost = {
  tags: Array<string>;
};

export type AllSanitySchemaTypes = BlogPost;"
`)
  })

  test('should generate correct types for document schema with unknown fields', () => {
    const schema: SchemaType = [
      {
        type: 'document',
        name: 'dynamicData',
        attributes: {
          metadata: {
            type: 'objectAttribute',
            value: {type: 'unknown'},
          },
        },
      },
    ]

    const typeGenerator = new TypeGenerator(schema)
    const actualOutput = typeGenerator.generateSchemaTypes()

    expect(actualOutput).toMatchInlineSnapshot(`
"export type DynamicData = {
  metadata: unknown;
};

export type AllSanitySchemaTypes = DynamicData;"
`)
  })

  test('should generate correct types for document schema with never fields', () => {
    const schema: SchemaType = [
      {
        type: 'document',
        name: 'impossible',
        attributes: {
          willNotHappen: {
            type: 'objectAttribute',
            value: {type: 'null'},
          },
        },
      },
    ]

    const typeGenerator = new TypeGenerator(schema)
    const actualOutput = typeGenerator.generateSchemaTypes()

    expect(actualOutput).toMatchInlineSnapshot(`
"export type Impossible = {
  willNotHappen: null;
};

export type AllSanitySchemaTypes = Impossible;"
`)
  })

  test('should generate correct types for document schema with custom type references', () => {
    const blogPost = {
      type: 'document',
      name: 'blogPost',
      attributes: {
        author: {
          type: 'objectAttribute',
          value: createReferenceTypeNode('author'),
        } satisfies ObjectAttribute<ObjectTypeNode>,
      },
    } satisfies DocumentSchemaType
    const author = {
      type: 'document',
      name: 'author',
      attributes: {
        name: {
          type: 'objectAttribute',
          value: {type: 'string'},
        } satisfies ObjectAttribute<StringTypeNode>,
      },
    } satisfies DocumentSchemaType
    const schema = [blogPost, author] satisfies SchemaType

    const typeGenerator = new TypeGenerator(schema)
    const actualOutput = typeGenerator.generateSchemaTypes()

    expect(actualOutput).toMatchInlineSnapshot(`
"export type BlogPost = {
  author: {
    _ref: string;
    _type: \\"reference\\";
    _weak?: boolean;
    [internalGroqTypeReferenceTo]?: \\"author\\";
  };
};

export type Author = {
  name: string;
};

export type AllSanitySchemaTypes = BlogPost | Author;"
`)
  })

  test('should generate correct types for document schema with union fields', () => {
    const schema: SchemaType = [
      {
        type: 'document',
        name: 'mixedContent',
        attributes: {
          content: {
            type: 'objectAttribute',
            value: {
              type: 'union',
              of: [{type: 'string'}, {type: 'number'}],
            },
          },
        },
      },
    ]

    const typeGenerator = new TypeGenerator(schema)
    const actualOutput = typeGenerator.generateSchemaTypes()

    expect(actualOutput).toMatchInlineSnapshot(`
"export type MixedContent = {
  content: string | number;
};

export type AllSanitySchemaTypes = MixedContent;"
`)
  })

  test('should generate correct types for document schema with nullable attribute', () => {
    const schema: SchemaType = [
      {
        type: 'document',
        name: 'optionalData',
        attributes: {
          obsoleteField: {
            type: 'objectAttribute',
            value: {type: 'null'},
          },
        },
      },
    ]

    const typeGenerator = new TypeGenerator(schema)
    const actualOutput = typeGenerator.generateSchemaTypes()

    expect(actualOutput).toMatchInlineSnapshot(`
"export type OptionalData = {
  obsoleteField: null;
};

export type AllSanitySchemaTypes = OptionalData;"
`)
  })

  describe('generateTypeNodeTypes', () => {
    const typeGenerator = new TypeGenerator([])
    test.each(['string', 'boolean', 'number', 'unknown', 'null'] as const)(
      'should be able to generate types for type nodes: %s',
      (typeName) => {
        const out = typeGenerator.generateTypeNodeTypes('test', {
          type: typeName,
        } satisfies TypeNode)

        expect(out).toMatchSnapshot()
      },
    )
  })

  test('should generate correct types for document schema with inline fields', () => {
    const objectNode = {
      type: 'object',
      attributes: {
        inlineField: {
          type: 'objectAttribute',
          value: {
            type: 'object',
            attributes: {
              test: {
                type: 'objectAttribute',
                value: {type: 'string'},
              },
            },
            rest: {
              type: 'inline',
              name: 'test',
            },
          },
        },
        unknownObject: {
          type: 'objectAttribute',
          value: {
            type: 'object',
            attributes: {
              test: {
                type: 'objectAttribute',
                value: {type: 'string'},
              },
            },
            rest: {
              type: 'unknown',
            },
          },
        },
        arrayField: {
          type: 'objectAttribute',
          value: {
            type: 'array',
            of: {type: 'string'},
          },
        },
        unionField: {
          type: 'objectAttribute',
          value: {
            type: 'union',
            of: [
              {
                type: 'object',
                attributes: {
                  test: {
                    type: 'objectAttribute',
                    value: {type: 'string'},
                  },
                },
              },
              {type: 'string'},
              {
                type: 'inline',
                name: 'test',
              },
            ],
          },
        },
      },
    } satisfies TypeNode

    const typeGenerator = new TypeGenerator([
      {
        type: 'type',
        name: 'test',
        value: {
          type: 'object',
          attributes: {test: {type: 'objectAttribute', value: {type: 'string'}}},
        },
      },
    ])
    const objectNodeOut = typeGenerator.generateTypeNodeTypes('myObject', objectNode)
    expect(objectNodeOut).toMatchSnapshot()

    const someOtherTypeOut = typeGenerator.generateTypeNodeTypes('SomeOtherType', {
      type: 'inline',
      name: 'myObject',
    })
    expect(someOtherTypeOut).toMatchSnapshot()
  })

  test('Adds a comment when missing referenced inline type', () => {
    const objectNode = {
      type: 'object',
      attributes: {
        test: {
          type: 'objectAttribute',
          value: {type: 'string'},
        },
      },
      rest: {
        type: 'inline',
        name: 'test',
      },
    } satisfies TypeNode

    const typeGenerator = new TypeGenerator([])
    const objectNodeOut = typeGenerator.generateTypeNodeTypes('myObject', objectNode)
    expect(objectNodeOut).toMatchSnapshot()
  })
})

describe('generateQueryMap', () => {
  test('should generate a map of query results', () => {
    const schema: SchemaType = []

    const queries = [
      {
        typeNode: {type: 'unknown'} satisfies TypeNode,
        query: '*[_type == "author"]',
      },
      {
        typeNode: {type: 'unknown'} satisfies TypeNode,
        query: '*[_type == "author"][0]',
      },
    ]

    const typeGenerator = new TypeGenerator(schema)
    typeGenerator.generateTypeNodeTypes('AuthorsResult', queries[0].typeNode)
    typeGenerator.generateTypeNodeTypes('FirstAuthorResult', queries[1].typeNode)

    const actualOutput = typeGenerator.generateQueryMap(queries)

    expect(actualOutput).toMatchInlineSnapshot(`
"import \\"@sanity/client\\";
declare module \\"@sanity/client\\" {
  interface SanityQueries {
    \\"*[_type == \\\\\\"author\\\\\\"]\\": AuthorsResult;
    \\"*[_type == \\\\\\"author\\\\\\"][0]\\": FirstAuthorResult;
  }
}"
`)
  })

  test('should generate a map of query results with duplicate type names', () => {
    const schema: SchemaType = []

    const queries = [
      {
        typeNode: {type: 'unknown'} satisfies TypeNode,
        query: '*[_type == "foo"]',
      },
      {
        typeNode: {type: 'unknown'} satisfies TypeNode,
        query: '*[_type == "bar"]',
      },
    ]

    const typeGenerator = new TypeGenerator(schema)
    typeGenerator.generateTypeNodeTypes('Foo', queries[0].typeNode)
    typeGenerator.generateTypeNodeTypes('Foo', queries[1].typeNode)

    const actualOutput = typeGenerator.generateQueryMap(queries)

    expect(actualOutput).toMatchInlineSnapshot(`
"import \\"@sanity/client\\";
declare module \\"@sanity/client\\" {
  interface SanityQueries {
    \\"*[_type == \\\\\\"foo\\\\\\"]\\": Foo;
    \\"*[_type == \\\\\\"bar\\\\\\"]\\": Foo_2;
  }
}"
`)
  })

  test('should generate a map of query results with duplicate query strings', () => {
    const schema: SchemaType = []

    const queries = [
      {
        typeNode: {type: 'unknown'} satisfies TypeNode,
        query: '*[_type == "foo"]',
      },
      {
        typeNode: {type: 'unknown'} satisfies TypeNode,
        query: '*[_type == "foo"]',
      },
    ]

    const typeGenerator = new TypeGenerator(schema)
    typeGenerator.generateTypeNodeTypes('Foo', queries[0].typeNode)
    typeGenerator.generateTypeNodeTypes('Bar', queries[1].typeNode)

    const actualOutput = typeGenerator.generateQueryMap(queries)

    expect(actualOutput).toMatchInlineSnapshot(`
"import \\"@sanity/client\\";
declare module \\"@sanity/client\\" {
  interface SanityQueries {
    \\"*[_type == \\\\\\"foo\\\\\\"]\\": Foo | Bar;
  }
}"
`)
  })
})
