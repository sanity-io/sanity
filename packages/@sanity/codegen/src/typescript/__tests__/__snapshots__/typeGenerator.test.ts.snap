// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateSchemaTypes can generate well known types 1`] = `"export declare const internalGroqTypeReferenceTo: unique symbol;"`;

exports[`generateSchemaTypes generateTypeNodeTypes should be able to generate types for type nodes: boolean 1`] = `"export type Test_2 = boolean;"`;

exports[`generateSchemaTypes generateTypeNodeTypes should be able to generate types for type nodes: null 1`] = `"export type Test_5 = null;"`;

exports[`generateSchemaTypes generateTypeNodeTypes should be able to generate types for type nodes: number 1`] = `"export type Test_3 = number;"`;

exports[`generateSchemaTypes generateTypeNodeTypes should be able to generate types for type nodes: string 1`] = `"export type Test = string;"`;

exports[`generateSchemaTypes generateTypeNodeTypes should be able to generate types for type nodes: unknown 1`] = `"export type Test_4 = unknown;"`;

exports[`generateSchemaTypes should generate TypeScript type declarations for a schema 1`] = `
"export type Author = {
  _id: string;
  _type: \\"author\\";
  _createdAt: string;
  _updatedAt: string;
  _rev: string;
  name?: string;
};

export type Post = {
  _id: string;
  _type: \\"post\\";
  _createdAt: string;
  _updatedAt: string;
  _rev: string;
  title?: string;
  author?: {
    _ref: string;
    _weak?: boolean;
    [internalGroqTypeReferenceTo]?: \\"author\\";
  } | {
    _ref: string;
    _weak?: boolean;
    [internalGroqTypeReferenceTo]?: \\"ghost\\";
  };
  slug?: Slug;
  excerpt?: string;
  mainImage?: {
    _type: \\"image\\";
    asset: {
      _ref: string;
      _weak?: boolean;
      [internalGroqTypeReferenceTo]?: \\"sanity.imageAsset\\";
    };
    caption?: string;
    attribution?: string;
    hotspot?: {
      _type: \\"sanity.imageHotspot\\";
      x: number;
      y: number;
      height: number;
      width: number;
    };
    crop?: {
      _type: \\"sanity.imageCrop\\";
      top: number;
      bottom: number;
      left: number;
      right: number;
    };
  };
  body?: BlockContent;
};

export type Ghost = {
  _id: string;
  _type: \\"ghost\\";
  _createdAt: string;
  _updatedAt: string;
  _rev: string;
  name?: string;
};

export type BlockContent = Array<{
  _key: string;
  level?: number;
  style?: \\"normal\\" | \\"h1\\" | \\"h2\\" | \\"h3\\" | \\"h4\\" | \\"blockquote\\";
  listItem?: \\"bullet\\";
  children: Array<{
    _key: string;
    text: string;
    marks: Array<string | \\"strong\\" | \\"em\\">;
  }>;
  markDefs: Array<{
    href?: string;
  }>;
  _key: string;
}>;

export type SanityAssetSourceData = {
  name?: string;
  id?: string;
  url?: string;
  _type: \\"sanity.assetSourceData\\";
};

export type Slug = {
  current?: string;
  source?: string;
  _type: \\"slug\\";
};

export type Geopoint = {
  lat?: number;
  lng?: number;
  alt?: number;
  _type: \\"geopoint\\";
};

export type SanityImageAsset = {
  _id: string;
  _type: \\"sanity.imageAsset\\";
  _createdAt: string;
  _updatedAt: string;
  _rev: string;
  originalFilename?: string;
  label?: string;
  title?: string;
  description?: string;
  altText?: string;
  sha1hash?: string;
  extension?: string;
  mimeType?: string;
  size?: number;
  assetId?: string;
  uploadId?: string;
  path?: string;
  url?: string;
  metadata?: SanityImageMetadata;
  source?: SanityAssetSourceData;
};

export type SanityFileAsset = {
  _id: string;
  _type: \\"sanity.fileAsset\\";
  _createdAt: string;
  _updatedAt: string;
  _rev: string;
  originalFilename?: string;
  label?: string;
  title?: string;
  description?: string;
  altText?: string;
  sha1hash?: string;
  extension?: string;
  mimeType?: string;
  size?: number;
  assetId?: string;
  uploadId?: string;
  path?: string;
  url?: string;
  source?: SanityAssetSourceData;
};

export type SanityImageCrop = {
  top?: number;
  bottom?: number;
  left?: number;
  right?: number;
  _type: \\"sanity.imageCrop\\";
};

export type SanityImageHotspot = {
  x?: number;
  y?: number;
  height?: number;
  width?: number;
  _type: \\"sanity.imageHotspot\\";
};

export type SanityImageMetadata = {
  location?: Geopoint;
  dimensions?: SanityImageDimensions;
  palette?: SanityImagePalette;
  lqip?: string;
  blurHash?: string;
  hasAlpha?: boolean;
  isOpaque?: boolean;
  _type: \\"sanity.imageMetadata\\";
};

export type SanityImageDimensions = {
  height?: number;
  width?: number;
  aspectRatio?: number;
  _type: \\"sanity.imageDimensions\\";
};

export type SanityImagePalette = {
  darkMuted?: SanityImagePaletteSwatch;
  lightVibrant?: SanityImagePaletteSwatch;
  darkVibrant?: SanityImagePaletteSwatch;
  vibrant?: SanityImagePaletteSwatch;
  dominant?: SanityImagePaletteSwatch;
  lightMuted?: SanityImagePaletteSwatch;
  muted?: SanityImagePaletteSwatch;
  _type: \\"sanity.imagePalette\\";
};

export type SanityImagePaletteSwatch = {
  background?: string;
  foreground?: string;
  population?: number;
  title?: string;
  _type: \\"sanity.imagePaletteSwatch\\";
};

export type AllSanitySchemaTypes = Author | Post | Ghost | BlockContent | SanityAssetSourceData | Slug | Geopoint | SanityImageAsset | SanityFileAsset | SanityImageCrop | SanityImageHotspot | SanityImageMetadata | SanityImageDimensions | SanityImagePalette | SanityImagePaletteSwatch;"
`;

exports[`generateSchemaTypes should generate correct types for document schema with inline fields 1`] = `
"export type MyObject = {
  inlineField: {
    test: string;
  } & Test;
  unknownObject: unknown;
  arrayField: Array<string>;
  unionField: {
    test: string;
  } | string | Test;
};"
`;

exports[`generateSchemaTypes should generate correct types for document schema with inline fields 2`] = `"export type SomeOtherType = MyObject;"`;
